ZEngine/include/ZEngine/Helpers/IntrusivePtr.h

```diff
@@ -0,0 +1,255 @@
+#pragma once
+#include <atomic>
+#include <functional>
+
+namespace ZEngine::Helpers
+{
+    class RefCounted
+    {
+    public:
+        RefCounted() : m_count(1) {}
+
+        RefCounted(const RefCounted&)            = delete;
+        RefCounted& operator=(const RefCounted&) = delete;
+
+        void increment()
+        {
+            ++m_count;
+        }
+
+        void decrement()
+        {
+            if (--m_count == 0)
+            {
+                delete this;
+            }
+        }
+
+        uint64_t RefCount() const
+        {
+            return m_count.load();
+        }
+
+        virtual ~RefCounted() = default;
+
+    private:
+        std::atomic<uint64_t> m_count;
+    };
+    
+    template <typename T>
+    class IntrusivePtr
+    {
+
+    public:
+
+        IntrusivePtr(T* ptr = nullptr) : m_ptr(ptr)
+        {
+            if (m_ptr)
+            {
+                m_ptr->increment();
+            }
```

Comment:
Let move this logic into a static method defined in RefCounted
e.g  :
static void RefCounted::IncrementRefCount(...)
static void RefCounted::DecrementRefCount(...)